operations/last_revision.go

```diff
@@ -0,0 +1,285 @@
+
+func newLastRevisionBuildInfo(b model.APIBuild, buildTasks []model.APITask) lastRevisionBuildInfo {
+	numSuccessfulTasks := 0
+	for _, t := range buildTasks {
+		if utility.FromStringPtr(t.Status) == evergreen.TaskSucceeded {
+			numSuccessfulTasks++
+		}
+	}
+	return lastRevisionBuildInfo{
+		buildID:            utility.FromStringPtr(b.Id),
+		buildVariant:       utility.FromStringPtr(b.BuildVariant),
+		versionID:          utility.FromStringPtr(b.Version),
+		numTasks:           len(buildTasks),
+		numSuccessfulTasks: numSuccessfulTasks,
+	}
+}
+
+// successProportion calculates the proportion of successful tasks out of all
+// the tasks in the build.
+func (i *lastRevisionBuildInfo) successProportion() float64 {
+	return float64(i.numSuccessfulTasks) / float64(i.numTasks)
+}
+
+// lastRevisionCriteria defines the user criteria for selecting a suitable last
+// revision.
+type lastRevisionCriteria struct {
+	// project is the project ID or identifier.
+	project string
+	// buildVariantRegexp is a list of regular expressions of matching build
+	// variant names. This determines which particular build variants the
+	// criteria should apply to.
+	buildVariantRegexp []regexp.Regexp
+	// minSuccessProportion is a criterion for the minimum proportion of tasks
+	// in a matching build that must succeed.
+	minSuccessProportion float64
+}
+
+func newLastRevisionCriteria(project string, bvRegexpsAsStr []string, minSuccessProportion float64) (*lastRevisionCriteria, error) {
+	if len(bvRegexpsAsStr) == 0 {
+		return nil, errors.New("must specify at least one build variant regexp for criteria")
+	}
+	if minSuccessProportion < 0 || minSuccessProportion > 1 {
+		return nil, errors.New("minimum success proportion must be between 0 and 1 inclusive")
+	}
+
+	bvRegexps := make([]regexp.Regexp, 0, len(bvRegexpsAsStr))
+	for _, bvRegexpStr := range bvRegexpsAsStr {
+		bvRegexp, err := regexp.Compile(bvRegexpStr)
+		if err != nil {
+			return nil, errors.Wrapf(err, "compiling build variant regexp '%s'", bvRegexpStr)
+		}
+		bvRegexps = append(bvRegexps, *bvRegexp)
+	}
+
+	return &lastRevisionCriteria{
+		project:              project,
+		buildVariantRegexp:   bvRegexps,
+		minSuccessProportion: minSuccessProportion,
+	}, nil
+}
+
+// shouldApply returns whether the criteria applies to this build variant.
+func (c *lastRevisionCriteria) shouldApply(bv string) bool {
+	for _, bvRegexp := range c.buildVariantRegexp {
+		if bvRegexp.MatchString(bv) {
+			return true
+		}
+	}
+	return false
+}
+
+// check returns whether the the criteria applies to the build and if so, if it
+// passes all the criteria. This returns true if the criteria does not apply.
+func (c *lastRevisionCriteria) check(info lastRevisionBuildInfo) bool {
+	if !c.shouldApply(info.buildVariant) {
+		// The criteria does not apply to this build variant, so it
+		// automatically passes checks.
+		return true
+	}
+
+	if info.successProportion() < c.minSuccessProportion {
+		grip.Debug(message.Fields{
```

Comment:
Since this is running in the cli I don't think this will send to splunk or log to the terminal. (Same for other grip statements)

Reply:
It will if you specify `evergreen --level=debug`, which I think could be useful for people who want to see their CLI's progress going through the versions.